src/webgpu/shader/validation/expression/call/builtin/refract.spec.ts

```diff
@@ -0,0 +1,158 @@
+const builtin = 'refract';
+export const description = `
+Validation tests for the ${builtin}() builtin.
+`;
+
+import { makeTestGroup } from '../../../../../../common/framework/test_group.js';
+import { keysOf, objectsToRecord } from '../../../../../../common/util/data_tables.js';
+import { Type, kConvertableToFloatVectors, scalarTypeOf } from '../../../../../util/conversion.js';
+import { ShaderValidationTest } from '../../../shader_validation_test.js';
+
+import {
+  ConstantOrOverrideValueChecker,
+  fullRangeForType,
+  kConstantAndOverrideStages,
+  stageSupportsType,
+  validateConstOrOverrideBuiltinEval,
+} from './const_override_validation.js';
+
+export const g = makeTestGroup(ShaderValidationTest);
+
+const kValidArgumentTypes = objectsToRecord(kConvertableToFloatVectors);
+
+g.test('values')
+  .desc(
+    `
+Validates that constant evaluation and override evaluation of ${builtin}() only errors in cases
+where a the calculations result in a non-representable value for the given type.
+`
+  )
+  .params(u =>
+    u
+      .combine('stage', kConstantAndOverrideStages)
+      .combine('type', keysOf(kValidArgumentTypes))
+      .filter(u => stageSupportsType(u.stage, kValidArgumentTypes[u.type]))
+      .beginSubcases()
+      .expand('a', u => fullRangeForType(kValidArgumentTypes[u.type], 5))
+      .expand('b', u => fullRangeForType(kValidArgumentTypes[u.type], 5))
+      .expand('c', u => fullRangeForType(kValidArgumentTypes[u.type], 5))
+  )
+  .beforeAllSubcases(t => {
+    if (scalarTypeOf(kValidArgumentTypes[t.params.type]) === Type.f16) {
+      t.selectDeviceOrSkipTestCase('shader-f16');
+    }
+  })
+  .fn(t => {
+    const type = kValidArgumentTypes[t.params.type];
+    const scalarType = scalarTypeOf(kValidArgumentTypes[t.params.type]);
+    const vCheck = new ConstantOrOverrideValueChecker(t, scalarType);
+
+    // Refract equation:
+    //   let k = 1.0 - c * c * (1.0 - dot(b, a) * dot(b, a))
+    //   if (k < 0.0) { return vecN(0.0); }
+    //   return c * a - (c * dot(b, a) + sqrt(k)) * b
+    // Should be invalid if the calculations result in intermediate values that
+    // exceed the maximum representable float value for the given type.
+    const a = Number(t.params.a);
+    const b = Number(t.params.b);
+    const c = Number(t.params.c);
+
+    const b_dot_a = vCheck.checkedResult(b * a * type.width);
+    const b_dot_a_2 = vCheck.checkedResult(b_dot_a * b_dot_a);
+    const one_minus_b_dot_a_2 = vCheck.checkedResult(1.0 - b_dot_a_2);
+    const c2 = vCheck.checkedResult(c * c);
+    const c2_one_minus_b_dot_a_2 = vCheck.checkedResult(c2 * one_minus_b_dot_a_2);
+    const k = vCheck.checkedResult(1.0 - c2_one_minus_b_dot_a_2);
+
+    if (k >= 0) {
+      const ca = vCheck.checkedResult(c * a);
+      const cbda = vCheck.checkedResult(c * b_dot_a);
+      const sqrt_k = vCheck.checkedResult(Math.sqrt(k));
```

Comment:
Since there is a sqrt being performed here, you will probably want to put in a isNearZero check similar to distance, since you could potentially hit the same 1/inverseqqrt(x) problems

Reply:
Done.